File: bindings/csharp/Sop.Examples/ConcurrentTransactionsDemoClustered.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.IO;

namespace Sop.Examples
{
    public static class ConcurrentTransactionsDemo
    {
        public static void Run()
        {
            // Enable verbose logging to stderr for debugging
            Logger.Configure(LogLevel.Debug, "");

            // Initialize Redis for Clustered mode
            try 
            {
                Redis.Initialize("redis://localhost:6379");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to connect to Redis: {ex.Message}");
                return;
            }

            Console.WriteLine("--- Concurrent Transactions Demo (Clustered) ---");
            Console.WriteLine("Demonstrating multi-threaded access without client-side locks.");
            Console.WriteLine("SOP handles ACID transactions, conflict detection, and merging.");
            Console.WriteLine("This runs in Clustered mode (Redis required).");

            string dbPath = "data/concurrent_demo_clustered";

            using var ctx = new Context();
            var db = new Database(new DatabaseOptions
            {
                StoresFolders = new List<string> { dbPath },
                Type = (int)DatabaseType.Clustered
            });

            // 1. Setup: Create the B-Tree in a separate transaction first
            // IMPORTANT: Pre-seed the B-Tree with one item to establish the root node.
            // This prevents race conditions on the very first commit when multiple threads 
            // try to initialize an empty tree simultaneously.
            using (var trans = db.BeginTransaction(ctx))
            {
                var btree = db.NewBtree<int, string>(ctx, "concurrent_tree", trans);
                btree.Add(ctx, new Item<int, string> { Key = -1, Value = "Root Seed Item" });
                trans.Commit();
            }

            Console.WriteLine("Launching parallel tasks...");
            
            int threadCount = 20;
            int itemsPerThread = 200;
            var threads = new List<Thread>();

            // We use a barrier to start threads roughly at the same time
            using var barrier = new Barrier(threadCount);

            for (int i = 0; i < threadCount; i++)
            {
                int threadId = i;
                var thread = new Thread(() => 
                {
                    // Use the shared Context (ctx) like Python does, instead of creating one per thread.
                    // SOP Context is thread-safe for ID retrieval.
                    
                    barrier.SignalAndWait();
                    
                    // NOTE: No 'lock' statement used here.
                    // Each thread gets its own Transaction.
                    // SOP manages isolation.
                    // This allows concurrent threads to do transactions without lock mgmt.
                    
                    int retryCount = 0;
                    bool committed = false;
                    while (!committed && retryCount < 10)
                    {
                        try 
                        {
                            Console.WriteLine($"Thread {threadId} starting transaction...");
                            using var trans = db.BeginTransaction(ctx);
                            Console.WriteLine($"Thread {threadId} opening btree...");
                            var btree = db.OpenBtree<int, string>(ctx, "concurrent_tree", trans);

                            for (int j = 0; j < itemsPerThread; j++)
                            {
                                // Unique keys per thread ensures no conflicts -> "Data Merge"
                                // If keys overlapped, SOP would detect conflicts and the commit might fail 
                                // (requiring a retry loop in a real app).
                                int key = (threadId * itemsPerThread) + j;
                                Console.WriteLine($"Thread {threadId} adding key {key}...");
                                btree.Add(ctx, new Item<int, string> { Key = key, Value = $"Thread {threadId} - Item {j}" });
                            }

                            Console.WriteLine($"Thread {threadId} committing...");
                            trans.Commit();
                            committed = true;
                            Console.WriteLine($"Thread {threadId} committed successfully.");
                        }
                        catch (Exception ex)
                        {
                            retryCount++;
                            // Random backoff to reduce contention
                            int delay = new Random().Next(100, 500) * retryCount;
                            Console.WriteLine($"Thread {threadId} conflict detected (Retry {retryCount}): {ex.Message}");
                            Thread.Sleep(delay);
                        }
                    }
                    
                    if (!committed)
                        Console.WriteLine($"Thread {threadId} failed after retries.");
                });
                threads.Add(thread);
                thread.Start();
            }

            foreach (var t in threads) t.Join();

            // Verify
            using (var trans = db.BeginTransaction(ctx, TransactionMode.ForReading))
            {
                var btree = db.OpenBtree<int, string>(ctx, "concurrent_tree", trans);
                
                long count = 0;
                if (btree.First(ctx))
                {
                    do
                    {
                        count++;
                    } while (btree.Next(ctx));
                }

                // Expected count includes the seed item
                long expectedCount = (threadCount * itemsPerThread) + 1;
                Console.WriteLine($"Final Count: {count} (Expected: {expectedCount})");

                if (count == expectedCount)
                    Console.WriteLine("SUCCESS: All transactions merged correctly.");
                else
                    Console.WriteLine("FAILURE: Count mismatch.");
            }
            
            db.RemoveBtree(ctx, dbPath);
            Redis.Close();
        }
    }
}

File: bindings/java/src/main/java/com/sharedcode/sop/examples/ConcurrentTransactionsDemoClustered.java
package com.sharedcode.sop.examples;

import com.sharedcode.sop.*;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

public class ConcurrentTransactionsDemoClustered {

    public static void main(String[] args) {
        run();
    }

    public static void run() {
        final String storeName = "concurrent_tree";

        // Enable verbose logging to stderr for debugging
        try {
            Logger.configure(LogLevel.Warn, "");
        } catch (SopException e) {
            e.printStackTrace();
        }

        System.out.println("--- Concurrent Transactions Demo (Clustered) ---");
        System.out.println("Demonstrating multi-threaded access without client-side locks.");
        System.out.println("SOP handles ACID transactions, conflict detection, and merging.");
        System.out.println("This runs in Clustered mode (Redis required).");

        String dbPath = "sop_data_concurrent_clustered";
        // Clean up previous run
        deleteDirectory(new File(dbPath));

        try (Context ctx = new Context()) {
            DatabaseOptions dbOpts = new DatabaseOptions();
            dbOpts.stores_folders = Collections.singletonList(dbPath);
            dbOpts.type = DatabaseType.Clustered;
            dbOpts.redis_config = new DatabaseOptions.RedisConfig();
            dbOpts.redis_config.address = "localhost:6379";
            dbOpts.redis_config.db = 0;

            Database db = new Database(dbOpts);

            // 1. Setup: Create the B-Tree in a separate transaction first
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<Integer, String> btree = db.newBtree(ctx, storeName, trans, null, Integer.class, String.class);
                btree.add(new Item<>( -1, "Root Seed Item"));
                trans.commit();
            } catch (Exception e) {
                e.printStackTrace();
                return;
            }

            System.out.println("Launching parallel tasks...");

            int threadCount = 20;
            int itemsPerThread = 200;
            List<Thread> threads = new ArrayList<>();
            Random rnd = new Random();

            for (int i = 0; i < threadCount; i++) {
                final int threadId = i;
                Thread thread = new Thread(() -> {
                    int retryCount = 0;
                    boolean committed = false;
                    while (!committed && retryCount < 2) {
                        try {
                            System.out.println("Thread " + threadId + " starting transaction...");
                            try (Transaction trans = db.beginTransaction(ctx)) {
                                System.out.println("Thread " + threadId + " opening btree...");

                                List<Item<Integer, String>> batch = new ArrayList<>(itemsPerThread);
                                BTree<Integer, String> btree = db.openBtree(ctx, storeName, trans, Integer.class, String.class);
                                
                                for (int j = 0; j < itemsPerThread; j++) {
                                    int key = (threadId * itemsPerThread) + j;
                                    batch.add(new Item<>(key, "Thread " + threadId + " - Item " + j));
                                }
                                
                                if (!btree.add(batch)) {
                                    System.out.println("Thread " + threadId + " failed to write batch");
                                }

                                System.out.println("Thread " + threadId + " committing...");
                                trans.commit();
                                committed = true;
                                System.out.println("Thread " + threadId + " committed successfully.");
                            }
                        } catch (Exception ex) {
                            retryCount++;
                            int delay = new Random().nextInt(400) + 100 * retryCount;
                            System.out.println("Thread " + threadId + " conflict detected (Retry " + retryCount + "): " + ex.getMessage());
                            try {
                                Thread.sleep(delay);
                            } catch (InterruptedException ie) {
                                Thread.currentThread().interrupt();
                            }
                        }
                    }

                    if (!committed) {
                        System.out.println("Thread " + threadId + " failed after retries.");
                    }
                });
                threads.add(thread);
                thread.start();

                try {
                    int delay = rnd.nextInt(480) + 20;
                    System.out.println("Waiting " + delay + "ms before starting next thread...");
                    Thread.sleep(delay);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            for (Thread t : threads) {
                try {
                    t.join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            // Verify
            try (Transaction trans = db.beginTransaction(ctx, TransactionMode.ForReading)) {
                BTree<Integer, String> btree = BTree.open(ctx, storeName, trans, Integer.class, String.class);

                long count = 0;
                if (btree.first()) {
                    do {
                        count++;
                    } while (btree.next());
                }

                long expectedCount = (threadCount * itemsPerThread) + 1;
                System.out.println("Final Count: " + count + " (Expected: " + expectedCount + ")");

                if (count == expectedCount) {
                    System.out.println("SUCCESS: All transactions merged correctly.");
                } else {
                    System.out.println("FAILURE: Count mismatch.");
                }
            } catch (Exception e) {
                e.printStackTrace();
            }

            // Cleanup
            try {
                db.removeBtree(ctx, storeName);
                Redis.close();
            } catch (Exception e) {
                e.printStackTrace();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void deleteDirectory(File file) {
        if (file.exists()) {
            File[] files = file.listFiles();
            if (files != null) {
                for (File f : files) {
                    deleteDirectory(f);
                }
            }
            file.delete();
        }
    }
}
